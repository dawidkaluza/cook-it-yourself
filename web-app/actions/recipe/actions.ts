"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
