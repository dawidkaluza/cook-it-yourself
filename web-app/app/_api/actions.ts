"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}