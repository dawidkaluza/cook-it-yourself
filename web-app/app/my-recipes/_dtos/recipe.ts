export type Recipe = {
  id?: number;
  name: string;
  description: string;
};

export type Ingredient = {
  id?: number;
  name: string;
  value: string;
  measure: string;
};

export type NewIngredient = Omit<Ingredient, "id">;
export type PersistedIngredient = Ingredient & { id: number };

export type Step = {
  id?: number;
  text: string;
};

export type NewStep = Omit<Step, "id">;
export type PersistedStep = Step & { id: number };

export type PortionSize = {
  value: string;
  measure: string;
};

export type RecipeDetails = Recipe & {
  ingredients: Ingredient[];
  methodSteps: Step[];
  cookingTime: number;
  portionSize: PortionSize;
};

export type UpdateRecipeRequest = {
  basicInformation?: {
    name: string;
    description: string;
    cookingTime: number;
    portionSize: PortionSize;
  };
  ingredients?: {
    ingredientsToAdd: NewIngredient[],
    ingredientsToUpdate: PersistedIngredient[],
    ingredientsToDelete: number[],
  };
  steps?: {
    stepsToAdd: NewStep[],
    stepsToUpdate: PersistedStep[],
    stepsToDelete: number[],
  };
};