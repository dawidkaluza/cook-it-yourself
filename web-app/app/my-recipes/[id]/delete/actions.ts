"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}