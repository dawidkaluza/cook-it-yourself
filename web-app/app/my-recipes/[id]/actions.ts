"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}