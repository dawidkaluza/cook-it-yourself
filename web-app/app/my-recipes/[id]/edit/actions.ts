"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };