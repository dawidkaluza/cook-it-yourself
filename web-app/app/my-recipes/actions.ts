"use server";

import {fetchFromServer} from "@/app/_api/fetch";
import {Page} from "@/app/my-recipes/_dtos/page";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";

export async function getMyRecipes(){
  return await fetchFromServer({ endpoint: "/kitchen/recipe" }) as Promise<Page<Recipe>>;
}