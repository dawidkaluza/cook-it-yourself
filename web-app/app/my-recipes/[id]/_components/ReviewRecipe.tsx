import {reviewRecipe} from "@/app/my-recipes/[id]/actions";
import Link from "next/link";

type Props = {
  id: number;
};

const ReviewRecipe = async ({ id } : Props) => {
  const recipe = await reviewRecipe(id);

  return (
    <div className="container-fluid">
      <div className="row">
        <div className="col-md-8 offset-md-2">
          <div>
            <h1>{recipe.name}</h1>
            <h6>{recipe.description}</h6>
            <ul className="list-unstyled">
              <li>Portion size: {Number(recipe.portionSize.value) + " " + recipe.portionSize.measure}</li>
              <li>Cooking time: {recipe.cookingTime} minutes</li>
            </ul>
          </div>

          <div className="mt-4">
            <h2>Ingredients</h2>
            <ul>
              {recipe.ingredients.map((ingredient) => (
                <li key={ingredient.id}>
                  {ingredient.name + " " + Number(ingredient.value) + " " + ingredient.measure}
                </li>
              ))}
            </ul>
          </div>

          <div className="mt-4">
            <h2>Method steps</h2>

            {recipe.methodSteps.map(step => (
              <p key={step.id}>{step.text}</p>
            ))}
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col-md-8 offset-md-2">
          <hr/>
        </div>

        {/* TODO test this part */}
        <div className="col-md-8 offset-md-2 d-flex justify-content-center justify-content-sm-start">
          <Link href={`/my-recipes/${recipe.id}/edit`} className="btn btn-primary mx-2">Modify</Link>
          <Link href={`/my-recipes/${recipe.id}/delete`} className="btn btn-danger mx-2">Delete</Link>
        </div>
      </div>
    </div>
  );
};

export {ReviewRecipe};