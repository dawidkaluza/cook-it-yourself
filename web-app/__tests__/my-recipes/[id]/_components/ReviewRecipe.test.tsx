import {Mock} from "vitest";
import {reviewRecipe} from "@/app/my-recipes/[id]/actions";
import {render, screen} from "@testing-library/react";
import {ReviewRecipe} from "@/app/my-recipes/[id]/_components/ReviewRecipe";
import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";

vi.mock("@/app/my-recipes/[id]/actions", () => {
  return {
    reviewRecipe: vi.fn(),
  };
});

describe("ReviewRecipe component", () => {
  test("renders given recipe", async () => {
    // Given
    const recipe : RecipeDetails = {
      id: 1,
      name: "Boiled sausages",
      description: "How to boil delicious sausages",
      ingredients: [
        {
          id: 1,
          name: "Sausage",
          value: "3",
          measure: "pc",
        }
      ],
      methodSteps: [
        {
          id: 1,
          text: "Boil sausages for about 3 mins",
        }
      ],
      cookingTime: 3,
      portionSize: {
        value: "1",
        measure: "plate",
      }
    };
    (reviewRecipe as Mock).mockResolvedValue(recipe);

    // When
    const reviewComponent = render(await ReviewRecipe({ id: 1 }));

    // Then
    const nameHeader = screen.getByRole("heading", { name: /Boiled sausages/ });
    expect(nameHeader).not.toBeNull();

    const descHeader = screen.getByRole("heading", { name: /How to boil delicious sausages/ });
    expect(descHeader).not.toBeNull();

    const portionSizeElement = screen.getByText("Portion size: " + Number(recipe.portionSize.value) + " " + recipe.portionSize.measure);
    expect(portionSizeElement).not.toBeNull();

    const cookingTimeElement = screen.getByText("Cooking time: " + recipe.cookingTime + " minutes");
    expect(cookingTimeElement).not.toBeNull();

    for(const ingredient of recipe.ingredients) {
      const ingredientElement = screen.getByText(
        ingredient.name + " " + Number(ingredient.value) + " " + ingredient.measure
      );

      expect(ingredientElement).not.toBeNull();
    }

    for(const step of recipe.methodSteps) {
      const stepElement = screen.getByText(step.text);
      expect(stepElement).not.toBeNull();
    }

    expect(reviewComponent.container).toMatchSnapshot();
  });
});