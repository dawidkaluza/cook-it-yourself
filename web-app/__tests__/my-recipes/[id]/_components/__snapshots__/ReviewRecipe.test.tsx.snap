// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewRecipe component > renders given recipe 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-md-8 offset-md-2"
      >
        <div>
          <h1>
            Boiled sausages
          </h1>
          <h6>
            How to boil delicious sausages
          </h6>
          <ul
            class="list-unstyled"
          >
            <li>
              Portion size: 
              1 plate
            </li>
            <li>
              Cooking time: 
              3
               minutes
            </li>
          </ul>
        </div>
        <div
          class="mt-4"
        >
          <h2>
            Ingredients
          </h2>
          <ul>
            <li>
              Sausage 3 pc
            </li>
          </ul>
        </div>
        <div
          class="mt-4"
        >
          <h2>
            Method steps
          </h2>
          <p>
            Boil sausages for about 3 mins
          </p>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-8 offset-md-2"
      >
        <hr />
      </div>
      <div
        class="col-md-8 offset-md-2 d-flex justify-content-center justify-content-sm-start"
      >
        <a
          class="btn btn-primary mx-2"
          href="/my-recipes/1/edit"
        >
          Modify
        </a>
        <a
          class="btn btn-danger mx-2"
          href="/my-recipes/1/delete"
        >
          Delete
        </a>
      </div>
    </div>
  </div>
</div>
`;
